File: DSA/SAMAGRA_590014891/Day_65_Question_2.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int first, second;
} Pair;

int comparePairs(const void *a, const void *b) {
    Pair *p1 = (Pair *)a;
    Pair *p2 = (Pair *)b;
    return p1->second - p2->second;
}

int maxChainLength(Pair pairs[], int n) {
    qsort(pairs, n, sizeof(Pair), comparePairs);
    int count = 1, last = pairs[0].second;
    for (int i = 1; i < n; i++) {
        if (pairs[i].first > last) {
            count++;
            last = pairs[i].second;
        }
    }
    return count;
}

int main() {
    int n;
    printf("Enter number of pairs: ");
    scanf("%d", &n);
    Pair pairs[n];
    printf("Enter pairs (a b):\n");
    for (int i = 0; i < n; i++) scanf("%d %d", &pairs[i].first, &pairs[i].second);
    printf("Maximum chain length: %d\n", maxChainLength(pairs, n));
    return 0;
}

File: DSA/AnshulJain_590014934/Day_65_Question_2.py
n = int(input("Enter number of pairs: "))
pairs = []
for _ in range(n):
    a, b = map(int, input().split())
    pairs.append((a, b))
pairs.sort(key=lambda x: x[1])
count = 0
end = float('-inf')
for a, b in pairs:
    if a > end:
        count += 1
        end = b
print(count)
